File: src/platforms/iacr.ts
import { mkdir, writeFile } from "node:fs/promises";
import { parse, type HTMLElement } from "node-html-parser/dist/index.js";

import type { Paper } from "../types.js";

const IACR_SEARCH_URL = "https://eprint.iacr.org/search";
const IACR_BASE_URL = "https://eprint.iacr.org";
const USER_AGENTS = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
];

function envInt(name: string, fallback: number): number {
  const raw = process.env[name];
  if (raw === undefined || raw.trim() === "") return fallback;
  const value = Number(raw);
  return Number.isFinite(value) && value > 0 ? value : fallback;
}

// Per-request timeouts. Defaults are generous so a slow proxy (SPIDER_PROXY)
// doesn't trip them on IACR's many sequential per-result requests; override via
// env when needed.
const REQUEST_TIMEOUT_MS = envInt("IACR_TIMEOUT_MS", 30_000);
const DOWNLOAD_TIMEOUT_MS = envInt("IACR_DOWNLOAD_TIMEOUT_MS", 60_000);

function getRandomUserAgent(): string {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)] ?? USER_AGENTS[0]!;
}

// eprint.iacr.org sits behind Cloudflare, which intermittently serves a JS
// "Just a moment…" challenge (HTTP 403/503 with an HTML body and a
// `cf-mitigated: challenge` header). Plain HTTP clients can't solve it — no
// header, cookie, retry, or proxy IP rotation gets past it — so we detect it and
// surface an actionable message instead of a bare "HTTP 403".
export function isCloudflareChallenge(response: Response): boolean {
  if (response.headers.get("cf-mitigated") === "challenge") return true;
  const server = response.headers.get("server") ?? "";
  const contentType = response.headers.get("content-type") ?? "";
  return (
    (response.status === 403 || response.status === 503) &&
    /cloudflare/i.test(server) &&
    contentType.includes("text/html")
  );
}

const CLOUDFLARE_NOTE =
  "eprint.iacr.org returned a Cloudflare bot challenge that automated clients cannot pass " +
  "(this is not a proxy or credential problem). Open the link in a browser to download it manually";

function createDocument(html: string): HTMLElement {
  return parse(html);
}

function parseDate(value?: string | null): string | null {
  if (!value) {
    return null;
  }

  const parsed = new Date(value.trim());
  return Number.isNaN(parsed.valueOf()) ? null : parsed.toISOString();
}

function textContent(element: HTMLElement | null): string {
  return element?.text.trim() ?? "";
}

export function buildIacrSearchParams(query: string, yearMin?: number, yearMax?: number): URLSearchParams {
  const params = new URLSearchParams({ q: query });
  if (yearMin !== undefined) {
    params.set("revisedafter", String(yearMin));
  }
  if (yearMax !== undefined) {
    params.set("revisedbefore", String(yearMax));
  }
  return params;
}

function parseIacrSearchResultElement(element: HTMLElement): Paper | null {
  const header = element.querySelector("div.d-flex");
  const content = element.querySelector("div.ms-md-4");
  const paperLink = header?.querySelector("a.paperlink");
  if (!header || !content || !paperLink) {
    return null;
  }

  const paperId = textContent(paperLink);
  const href = paperLink.getAttribute("href") ?? "";
  const pdfLink = (Array.from(header.querySelectorAll("a")) as HTMLElement[]).find((link) => textContent(link) === "(PDF)");
  const updatedText = textContent(header.querySelector("small.ms-auto")).replace("Last updated:", "").trim();
  const updatedDate = parseDate(updatedText);

  return {
    paperId,
    title: textContent(content.querySelector("strong")),
    authors: textContent(content.querySelector("span.fst-italic"))
      .split(",")
      .map((author) => author.trim())
      .filter(Boolean),
    abstract: textContent(content.querySelector("p.search-abstract")),
    doi: "",
    publishedDate: updatedDate ?? new Date(Date.UTC(1900, 0, 1)).toISOString(),
    pdfUrl: pdfLink ? `${IACR_BASE_URL}${pdfLink.getAttribute("href") ?? ""}` : "",
    url: `${IACR_BASE_URL}${href}`,
    source: "iacr",
    updatedDate,
    categories: textContent(content.querySelector("small.badge")) ? [textContent(content.querySelector("small.badge"))] : [],
    keywords: [],
    citations: 0,
    references: [],
    extra: {},
  };
}

export function parseIacrSearchHtml(html: string, _fetchDetails = false): Paper[] {
  const document = createDocument(html);
  return Array.from(document.querySelectorAll("div.mb-4"))
    .map((element) => parseIacrSearchResultElement(element))
    .filter((paper): paper is Paper => paper !== null);
}

export function parseIacrDetailHtml(html: string, paperId: string): Paper | null {
  const document = createDocument(html);
  const lines = document.text
    .split("\n")
    .map((line: string) => line.trim())
    .filter(Boolean);

  let publicationInfo = "";
  let historyStarted = false;
  const historyEntries: string[] = [];
  let updatedDate: string | null = null;

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index]!;

    if (line === "Publication info" && index + 1 < lines.length) {
      publicationInfo = lines[index + 1] ?? "";
      continue;
    }

    if (line === "History") {
      historyStarted = true;
      continue;
    }

    if (historyStarted && (line.startsWith("Short URL") || line.startsWith("License"))) {
      break;
    }

    if (historyStarted && line.includes(":")) {
      historyEntries.push(line);
      if (!updatedDate) {
        updatedDate = parseDate(line.split(":")[0] ?? "");
      }
    }
  }

  const title = textContent(document.querySelector("h3.mb-3"));
  if (!title) {
    return null;
  }

  const authors = textContent(document.querySelector("p.fst-italic"))
    .replaceAll(" and ", ",")
    .split(",")
    .map((author) => author.trim())
    .filter(Boolean);

  return {
    paperId,
    title,
    authors,
    abstract: textContent(document.querySelector('p[style="white-space: pre-wrap;"]')),
    doi: "",
    publishedDate: updatedDate ?? new Date().toISOString(),
    pdfUrl: `${IACR_BASE_URL}/${paperId}.pdf`,
    url: `${IACR_BASE_URL}/${paperId}`,
    source: "iacr",
    updatedDate,
    categories: [],
    keywords: Array.from(document.querySelectorAll("a.badge.bg-secondary.keyword"))
      .map((element) => textContent(element))
      .filter(Boolean),
    citations: 0,
    references: [],
    extra: {
      publicationInfo,
      history: historyEntries.join("; "),
    },
  };
}

export async function getIacrPaperDetails(paperId: string): Promise<Paper | null> {
  const response = await fetch(`${IACR_BASE_URL}/${paperId}`, {
    headers: {
      "User-Agent": getRandomUserAgent(),
      Accept: "text/html,application/xhtml+xml",
      "Accept-Language": "en-US,en;q=0.9",
    },
    signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
  });

  if (!response.ok) {
    return null;
  }

  return parseIacrDetailHtml(await response.text(), paperId);
}

export async function searchIacrPapers(
  query: string,
  options: { maxResults?: number; fetchDetails?: boolean; yearMin?: number; yearMax?: number } = {},
): Promise<Paper[]> {
  const url = new URL(IACR_SEARCH_URL);
  url.search = buildIacrSearchParams(query, options.yearMin, options.yearMax).toString();

  const response = await fetch(url, {
    headers: {
      "User-Agent": getRandomUserAgent(),
      Accept: "text/html,application/xhtml+xml",
      "Accept-Language": "en-US,en;q=0.9",
    },
    signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
  });

  if (!response.ok) {
    if (isCloudflareChallenge(response)) {
      throw new Error(`${CLOUDFLARE_NOTE}: ${IACR_SEARCH_URL}`);
    }
    return [];
  }

  const fallbackResults = parseIacrSearchHtml(await response.text());
  const maxResults = options.maxResults ?? 10;
  const results: Paper[] = [];

  for (const paper of fallbackResults) {
    if (results.length >= maxResults) {
      break;
    }

    if (options.fetchDetails !== false) {
      const detailed = await getIacrPaperDetails(paper.paperId);
      results.push(detailed ?? paper);
    } else {
      results.push(paper);
    }
  }

  return results;
}

export async function downloadIacrPaperPdf(paperId: string, savePath: string): Promise<string> {
  await mkdir(savePath, { recursive: true });
  const url = `${IACR_BASE_URL}/${paperId}.pdf`;
  const response = await fetch(url, {
    headers: {
      "User-Agent": getRandomUserAgent(),
      Accept: "application/pdf,text/html;q=0.9,*/*;q=0.8",
      "Accept-Language": "en-US,en;q=0.9",
      Referer: `${IACR_BASE_URL}/${paperId}`,
    },
    redirect: "follow",
    signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS),
  });

  if (!response.ok) {
    if (isCloudflareChallenge(response)) {
      return `Failed to download PDF: ${CLOUDFLARE_NOTE}: ${url}`;
    }
    return `Failed to download PDF: HTTP ${response.status}`;
  }

  // A challenge page can also come back as 200 text/html — never save that as a PDF.
  const contentType = response.headers.get("content-type") ?? "";
  if (!contentType.includes("pdf")) {
    return `Failed to download PDF: expected a PDF but received ${contentType || "an unknown type"}. ${CLOUDFLARE_NOTE}: ${url}`;
  }

  const filename = `${savePath}/iacr_${paperId.replaceAll("/", "_")}.pdf`;
  await writeFile(filename, Buffer.from(await response.arrayBuffer()));
  return filename;
}

File: src/apaper_mcp/platforms/iacr.py
from datetime import datetime, timezone
import os
from pathlib import Path
import shutil
from urllib.parse import urlencode

from bs4 import BeautifulSoup
from scrapling.fetchers import StealthyFetcher

from ..proxy import selenium_proxy_from_env


def build_iacr_search_params(
    query: str, year_min: int | None = None, year_max: int | None = None
) -> str:
    values = {"q": query}
    if year_min is not None:
        values["revisedafter"] = str(year_min)
    if year_max is not None:
        values["revisedbefore"] = str(year_max)
    return urlencode(values)


def _scrapling_proxy(proxy: str | None) -> str | None:
    """Return the proxy format expected by Scrapling/Playwright."""
    if proxy is None or "://" in proxy:
        return proxy
    return f"http://{proxy}"


def download_iacr_pdf(pdf_url: str, target: Path, fetcher=None) -> None:
    """Use Scrapling's stealth browser to fetch and atomically save an IACR PDF."""
    fetcher = fetcher or StealthyFetcher
    timeout = int(os.getenv("IACR_DOWNLOAD_TIMEOUT_MS", "600000"))
    attempts = max(1, int(os.getenv("IACR_CAPTCHA_ATTEMPTS", "3")))
    detail_url = pdf_url.removesuffix(".pdf")

    fetch_options = {
        "solve_cloudflare": True,
        "timeout": timeout,
        "retries": 1,
        "locale": "en",
    }
    browser_executable = os.getenv("IACR_BROWSER_EXECUTABLE") or shutil.which(
        "chromium"
    )
    if browser_executable:
        fetch_options["executable_path"] = browser_executable

    body = b""
    last_error = None
    with selenium_proxy_from_env() as proxy:
        fetch_options["proxy"] = _scrapling_proxy(proxy)
        for _ in range(attempts):
            result = {}

            def request_pdf(page) -> None:
                try:
                    response = page.request.get(pdf_url, timeout=timeout)
                    result.update(
                        body=response.body(),
                        status=response.status,
                        content_type=response.headers.get("content-type", "unknown"),
                    )
                except Exception as error:
                    result["error"] = error

            try:
                fetcher.fetch(detail_url, page_action=request_pdf, **fetch_options)
            except Exception as error:
                last_error = error
                continue

            if result.get("error") is not None:
                last_error = result["error"]
                continue
            status = int(result.get("status", 0))
            if status < 200 or status >= 300:
                last_error = RuntimeError(
                    f"IACR returned HTTP {status or 'unknown'} for {pdf_url}"
                )
                continue
            body = result.get("body", b"")
            if body.startswith(b"%PDF"):
                break
            last_error = ValueError(
                "Scrapling did not receive a valid PDF "
                f"(content-type: {result.get('content_type', 'unknown')})"
            )
        else:
            if last_error is not None:
                raise last_error
            raise ValueError("Scrapling did not receive a valid PDF")

    target.parent.mkdir(parents=True, exist_ok=True)
    partial = target.with_name(target.name + ".part")
    partial.unlink(missing_ok=True)
    try:
        partial.write_bytes(body)
        partial.replace(target)
    finally:
        partial.unlink(missing_ok=True)


def _paper(**kwargs):
    return {
        "paper_id": "",
        "title": "",
        "authors": [],
        "abstract": "",
        "doi": "",
        "published_date": None,
        "pdf_url": "",
        "url": "",
        "source": "iacr",
        "categories": [],
        "keywords": [],
        "citations": 0,
        "references": [],
        "extra": {},
        **kwargs,
    }


def parse_iacr_search_html(html: str, _fetch_details: bool = False) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    papers = []
    for item in soup.select("div.mb-4"):
        header, content, link = (
            item.select_one("div.d-flex"),
            item.select_one("div.ms-md-4"),
            item.select_one("a.paperlink"),
        )
        if not header or not content or not link:
            continue
        updated = (
            header.select_one("small.ms-auto")
            .get_text(strip=True)
            .replace("Last updated:", "")
            .strip()
            if header.select_one("small.ms-auto")
            else ""
        )
        try:
            updated = (
                datetime.fromisoformat(updated)
                .replace(tzinfo=timezone.utc)
                .isoformat()
                .replace("+00:00", "Z")
            )
        except ValueError:
            updated = None
        pdf = next(
            (
                x.get("href")
                for x in header.select("a")
                if x.get_text(strip=True) == "(PDF)"
            ),
            "",
        )
        authors = [
            x.strip()
            for x in (
                content.select_one("span.fst-italic").get_text()
                if content.select_one("span.fst-italic")
                else ""
            ).split(",")
            if x.strip()
        ]
        papers.append(
            _paper(
                paper_id=link.get_text(strip=True),
                title=content.select_one("strong").get_text(strip=True)
                if content.select_one("strong")
                else "",
                authors=authors,
                abstract=content.select_one("p.search-abstract").get_text(strip=True)
                if content.select_one("p.search-abstract")
                else "",
                published_date=updated or "1900-01-01T00:00:00Z",
                updated_date=updated,
                pdf_url=f"https://eprint.iacr.org{pdf}",
                url=f"https://eprint.iacr.org{link.get('href', '')}",
                categories=[content.select_one("small.badge").get_text(strip=True)]
                if content.select_one("small.badge")
                else [],
            )
        )
    return papers


def parse_iacr_detail_html(html: str, paper_id: str) -> dict | None:
    soup = BeautifulSoup(html, "html.parser")
    title = soup.select_one("h3.mb-3")
    if not title:
        return None
    lines = [line.strip() for line in soup.get_text("\n").splitlines() if line.strip()]
    history = []
    publication = ""
    started = False
    updated = None
    for index, line in enumerate(lines):
        if line == "Publication info" and index + 1 < len(lines):
            publication = lines[index + 1]
        if line == "History":
            started = True
            continue
        if started and (line.startswith("Short URL") or line.startswith("License")):
            break
        if started and ":" in line:
            history.append(line)
            if updated is None:
                try:
                    updated = (
                        datetime.fromisoformat(line.split(":", 1)[0])
                        .replace(tzinfo=timezone.utc)
                        .isoformat()
                        .replace("+00:00", "Z")
                    )
                except ValueError:
                    pass
    authors = [
        x.strip()
        for x in (
            soup.select_one("p.fst-italic").get_text()
            if soup.select_one("p.fst-italic")
            else ""
        )
        .replace(" and ", ",")
        .split(",")
        if x.strip()
    ]
    return _paper(
        paper_id=paper_id,
        title=title.get_text(strip=True),
        authors=authors,
        abstract=soup.select_one('p[style="white-space: pre-wrap;"]').get_text(
            strip=True
        )
        if soup.select_one('p[style="white-space: pre-wrap;"]')
        else "",
        published_date=updated
        or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        updated_date=updated,
        pdf_url=f"https://eprint.iacr.org/{paper_id}.pdf",
        url=f"https://eprint.iacr.org/{paper_id}",
        keywords=[
            x.get_text(strip=True) for x in soup.select("a.badge.bg-secondary.keyword")
        ],
        extra={"publicationInfo": publication, "history": "; ".join(history)},
    )
